src/MOI_wrapper.jl

```diff
@@ -0,0 +1,414 @@
+    scip, cons = get_scip(o), get_cons(o, ci)
+    lhs, rhs = SCIPgetLhsLinear(scip, cons), SCIPgetRhsLinear(scip, cons)
+    from_bounds(S, lhs, rhs)
+end
+
+function MOI.get(o::Optimizer, ::MOI.ConstraintName, ci::CI{SAF,<:BOUNDS})
+    SCIPconsGetName(get_cons(o, ci))
+end
+
+function MOI.set(o::Optimizer, ::MOI.ConstraintName, ci::CI{SAF,<:BOUNDS}, name::String)
+    @SC SCIPchgConsName(get_scip(o), get_cons(o, ci), name)
+end
+
+function MOI.set(o::Optimizer, ::MOI.ObjectiveFunction{SAF}, obj::SAF)
+    allow_modification(o)
+    scip = get_scip(o)
+
+    # reset objective coefficient of all variables first
+    for v in o.mscip.vars
+        @SC SCIPchgVarObj(scip, v[], 0.0)
+    end
+
+    # set new objective coefficients, summing coefficients
+    for t in obj.terms
+        var = get_var(o, t.variable_index)
+        oldcoef = SCIPvarGetObj(var)
+        newcoef = oldcoef + t.coefficient
+        @SC SCIPchgVarObj(scip, var, newcoef)
+    end
+
+    @SC SCIPaddOrigObjoffset(scip, obj.constant - SCIPgetOrigObjoffset(scip))
+
+    return nothing
+end
+
+function MOI.get(o::Optimizer, ::MOI.ObjectiveFunction{SAF})
+    terms = SAT[]
+    for i = 1:length(o.mscip.vars)
+        vi = VI(i)
+        coef = SCIPvarGetObj(get_var(o, vi))
+        coef == 0.0 || push!(terms, SAT(coef, vi))
+    end
+    constant = SCIPgetOrigObjoffset(get_scip(o))
+    return SAF(terms, constant)
+end
+
+function MOI.set(o::Optimizer, ::MOI.ObjectiveSense, sense::MOI.OptimizationSense)
+    allow_modification(o)
+    if sense == MOI.MIN_SENSE
+        @SC SCIPsetObjsense(get_scip(o), SCIP_OBJSENSE_MINIMIZE)
+    elseif sense == MOI.MAX_SENSE
+        @SC SCIPsetObjsense(get_scip(o), SCIP_OBJSENSE_MAXIMIZE)
+    end
+    return nothing
+end
+
+function MOI.get(o::Optimizer, ::MOI.ObjectiveSense)
+    SCIPgetObjsense(get_scip(o)) == SCIP_OBJSENSE_MAXIMIZE ?
+        MOI.MAX_SENSE :
+        MOI.MIN_SENSE
+end
+
+function MOI.modify(o::Optimizer, ci::CI{SAF, <:BOUNDS}, change::SCC)
+    allow_modification(o)
+    @SC SCIPchgCoefLinear(get_scip(o), get_cons(o, ci),
+                          get_var(o, change.variable), change.new_coefficient)
+    return nothing
+end
+
+function MOI.modify(o::Optimizer, ::MOI.ObjectiveFunction{SAF}, change::SCC)
+    allow_modification(o)
+    @SC SCIPchgVarObj(get_scip(o), get_var(o, change.variable),
+                      change.new_coefficient)
+    return nothing
+end
+
+## optimization and results
+
+function MOI.optimize!(o::Optimizer)
+    @SC SCIPsolve(o.mscip.scip[])
+    return nothing
+end
+
+term_status_map = Dict(
+    SCIP_STATUS_UNKNOWN => MOI.OPTIMIZE_NOT_CALLED,
+    SCIP_STATUS_USERINTERRUPT => MOI.INTERRUPTED,
+    SCIP_STATUS_NODELIMIT => MOI.NODE_LIMIT,
+    SCIP_STATUS_TOTALNODELIMIT => MOI.NODE_LIMIT,
+    SCIP_STATUS_STALLNODELIMIT => MOI.OTHER_LIMIT,
+    SCIP_STATUS_TIMELIMIT => MOI.TIME_LIMIT,
+    SCIP_STATUS_MEMLIMIT => MOI.MEMORY_LIMIT,
+    SCIP_STATUS_GAPLIMIT => MOI.OTHER_LIMIT,
```

Comment:
Other solvers map this to `OPTIMAL`. I.e., terminating with a gap is optimal within the user-provided termination criteria.

Reply:
OK, done